File: spark-consumer/src/main/scala/io/acalio/utility/sentiment/SentimentAnalyzer.scala
// package io.acalio.utility.sentiment
// import edu.stanford.nlp.pipeline.Annotation
// import edu.stanford.nlp.ling.CoreAnnotations
// import edu.stanford.nlp.sentiment.SentimentCoreAnnotations
// import edu.stanford.nlp.trees.Tree;
// import edu.stanford.nlp.neural.rnn.RNNCoreAnnotations
// import edu.stanford.nlp.pipeline.StanfordCoreNLP
// import eu.fbk.dh.tint.runner.TintPipeline
// import java.{util => ju}
// import com.github.pemistahl.lingua.api._
// import com.github.pemistahl.lingua.api.Language._

// object SentimentAnalyzer {

//   @transient private var enSentimentPipeline : StanfordCoreNLP = _
//   @transient private var itSentimentPipeline : TintPipeline = _

//   // import Language.Language
//   // private var lan : Language = Language.EN

//   // def setLanguage(l: Language) {
//   //   lan = l
//   // }

//   def getLanguage(text: String): Language = {
//     val detector = LanguageDetectorBuilder
//       .fromLanguages(ENGLISH, ITALIAN)
//       .build
//     return detector.detectLanguageOf(text)
//   }

//   def sentiment(text:String): Option[Double] = {
//     var annotation: Annotation = null
//     val lan = getLanguage(text)
//     lan match {
//       case ENGLISH => {
//         annotation = sentimentEn(text)
//       }
//       case ITALIAN => {
//         annotation = sentimentIt(text)
//       }
//       case _ => return None
//     }

//     //val the the sentiment score of each sentence
//     var sentimentScore: Double = 0
//     annotation
//       .get(classOf[CoreAnnotations.SentencesAnnotation])
//       .forEach(s =>
//         sentimentScore+=RNNCoreAnnotations
//           .getPredictedClass(s.get(classOf[SentimentCoreAnnotations.SentimentAnnotatedTree])))
//     //the sentiment score is the average sentiment among every tree in the text
//     return Some(sentimentScore / annotation.get(classOf[CoreAnnotations.SentencesAnnotation]).size().toDouble)
//   }


//   def sentimentEn(text: String): Annotation = {
//     if (enSentimentPipeline == null) {
//       val props = new ju.Properties
//       props.setProperty("annotators", "tokenize, ssplit, parse, sentiment")
//       enSentimentPipeline = new StanfordCoreNLP(props)
//     }
//     // Get the original Annotation (Stanford CoreNLP)
//     return enSentimentPipeline.process(text);
//   }

//   def sentimentIt(text: String): Annotation = {
//     if (itSentimentPipeline == null) {
//       itSentimentPipeline = new TintPipeline
//       // Load the default properties
//       // see https://github.com/dhfbk/tint/blob/master/tint-runner/src/main/resources/default-config.properties
//       itSentimentPipeline.loadDefaultProperties();
//       // Add a custom property
//       itSentimentPipeline.setProperty("annotators", "tokenize, ssplit, pos, parse, sentiment");
//       // Load the models
//       itSentimentPipeline.load();
//     }
//     return itSentimentPipeline.runRaw(text)
//   }
  
// }

// // object Language extends Enumeration {
// //   type Language = Value
// //   val EN, IT = Value
// //   def get(l: String): Language = {
// //     l match {
// //       case "en" => return EN
// //       case "it" => return IT
// //       case _ => throw new RuntimeException(s"Unrecognized option: $l")
// //     }

// //   }
// // }

File: nlp/src/main/java/io/acalio/nlp/SentimentAnalyzer.java
package io.acalio.nlp;
import java.io.IOException;
import java.io.Serializable;
import java.util.Properties;

import com.github.pemistahl.lingua.api.*;

import edu.stanford.nlp.ling.CoreAnnotations;
import edu.stanford.nlp.neural.rnn.RNNCoreAnnotations;
import edu.stanford.nlp.pipeline.Annotation;
import edu.stanford.nlp.pipeline.StanfordCoreNLP;
import edu.stanford.nlp.sentiment.SentimentCoreAnnotations;
import edu.stanford.nlp.util.CoreMap;
import eu.fbk.dh.tint.runner.TintPipeline;


/**
 * This class provide the ability to detect a language 
 * of a raw text and retrieve its sentiment. 
 * Even though we are able to detect several langauges,
 * only the sentiment for italian and english texts can be extracted 
 *
 */
public class SentimentAnalyzer implements Serializable {

    private transient static TintPipeline itPipeline = new TintPipeline();
    private transient static StanfordCoreNLP enPipeline = null;
    private transient static LanguageDetector detector = null;

    static {
	try {
	    //initialize the pipeline for the italian language
	    itPipeline.loadDefaultProperties();
	    itPipeline.setProperty("annotators", "tokenize, ssplit, pos, parse, sentiment");
	    itPipeline.load();
	    
	    //initialize core nlp
	    Properties props = new Properties();
	    props.setProperty("annotators", "tokenize, ssplit, parse, sentiment");
	    enPipeline = new StanfordCoreNLP(props);
	    
	    //initialize language detector
	    detector = LanguageDetectorBuilder
		.fromLanguages(Language.ENGLISH,
			       Language.ITALIAN,
			       Language.GERMAN,
			       Language.FRENCH,
			       Language.SPANISH).build();
	    
	} catch (IOException e) {
		e.printStackTrace();
	}

    }


    

    /**
     * Get the sentiment associated with the text
     * provided as input. 
     * The sentiment is given by the average sentiment
     * among all the senteces in the text
     *
     * @param text: String
     * @return  the text sentiment
     */ 
    public static double getSentiment(String text) {
	Language language = detector.detectLanguageOf(text);
	Annotation annotations = null;
	switch (language) {
	case ITALIAN:
	    annotations = itPipeline.runRaw(text);
	    break;
	case ENGLISH:
	    annotations = enPipeline.process(text);
	    break;
	default:
	    //other languages are ignored
	    return Double.MAX_VALUE;
	}
	double sentiment = 0;
	for(CoreMap sentence : annotations.get(CoreAnnotations.SentencesAnnotation.class))
	    sentiment += RNNCoreAnnotations
		.getPredictedClass(
		   sentence.get(SentimentCoreAnnotations.SentimentAnnotatedTree.class)		      
		);

	sentiment /= (double) annotations.get(CoreAnnotations.SentencesAnnotation.class).size();
	return sentiment;
    }

    /**
     * Get the language of the text proivded as input
     * 
     * @param text
     * @return the text language
     *
     */
    public static Language getLanguage(String text) {
	return detector.detectLanguageOf(text);
    }

    

}
